synapseclient/api/entity_services.py

```diff
@@ -964,3 +1245,305 @@ async def main():
+                entity_id="syn123",
+                principal_id="12345"
+            )
+            print(f"User 12345 permissions: {permissions}")
+
+        asyncio.run(main())
+        ```
+    """
+    from synapseclient import AUTHENTICATED_USERS, PUBLIC
+    from synapseclient.api.team_services import get_teams_for_user
+    from synapseclient.api.user_services import (
+        get_user_bundle,
+        get_user_by_principal_id_or_name,
+    )
+
+    # Get the ACL for the entity
+    acl = await get_entity_acl_with_benefactor(
+        entity_id=entity_id,
+        check_benefactor=check_benefactor,
+        synapse_client=synapse_client,
+    )
+
+    # Get the principal ID as an integer (None defaults to PUBLIC)
+    principal_id_int = await get_user_by_principal_id_or_name(
+        principal_id=principal_id, synapse_client=synapse_client
+    )
+
+    # Get teams that the user belongs to
+    team_ids = []
+    async for team in get_teams_for_user(
+        user_id=str(principal_id_int), synapse_client=synapse_client
+    ):
+        team_ids.append(int(team["id"]))
+
+    user_profile_bundle = await get_user_bundle(
+        user_id=principal_id_int, mask=1, synapse_client=synapse_client
+    )
+
+    effective_permission_set = set()
+
+    # Loop over all permissions in the returned ACL and add it to the effective_permission_set
+    # if the principalId in the ACL matches
+    # 1) the one we are looking for,
+    # 2) a team the user is a member of,
+    # 3) PUBLIC
+    # 4) AUTHENTICATED_USERS (if user_profile_bundle exists for the principal_id)
+    for permissions in acl["resourceAccess"]:
+        if "principalId" in permissions and (
+            permissions["principalId"] == principal_id_int
+            or permissions["principalId"] in team_ids
+            or permissions["principalId"] == PUBLIC
+            or (
+                permissions["principalId"] == AUTHENTICATED_USERS
+                and user_profile_bundle is not None
+            )
+        ):
+            effective_permission_set = effective_permission_set.union(
+                permissions["accessType"]
+            )
+    return list(effective_permission_set)
+
+
+async def update_entity_acl(
+    entity_id: str,
+    acl: Dict[str, Any],
+    *,
+    synapse_client: Optional["Synapse"] = None,
+) -> Dict[str, Any]:
+    """
+    Create or update the Access Control List(ACL) for an entity.
+
+    Arguments:
+        entity_id: The ID of the entity.
+        acl: The ACL to be applied to the entity. Should match the format:
+            {'resourceAccess': [
+                {'accessType': ['READ', 'DOWNLOAD'],
+                 'principalId': 222222}
+            ]}
+        synapse_client: If not passed in and caching was not disabled by
+                `Synapse.allow_client_caching(False)` this will use the last created
+                instance from the Synapse class constructor.
+
+    Returns:
+        The created or updated ACL.
```

Comment:
Nit: add link to ACL object: https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/AccessControlList.html